File: ch4/listing4.2.py
## Yi-Shin Lin, 16-11-2021
import numpy as np
from scipy.stats import binom
import scipy.optimize
import pandas as pd

## Obtain current predictions and compute discrepancy
def Euclidean(x, y, w):
    ## Euclidean(stim[:1].to_numpy(), stim[:2].to_numpy(), w)
    term0 = w*np.power(x - y, 2)
    out = np.sqrt( np.sum(term0) )
    return out

def GCMpred(probe, exemplars, c, w):
    ## tmp = GCMpred(stim[:1].to_numpy(), exemplars, c, w)
    ## {0: array([0.        , 0.77503833, 0.41018329, 0.54557113, 0.82417462]),
    ##  1: array([0.87030496, 1.01944575, 0.80819033, 1.24746426, 1.33244921])}
    nexem = len(exemplars)
    out = np.empty(nexem)
    for i, key in enumerate(exemplars.keys()):
        ex = exemplars[key]
        nobs = len(ex)
        
        tmp = np.empty(nobs)
        tmp[:] = np.nan
        for j in range(0, nobs):
            tmp[j] = Euclidean(ex.iloc[j].to_numpy(), probe, w)

        out[i] = np.sum(np.exp(-c*tmp))

    return out / np.sum(out)

N = 2*80 ## two responses per face from 80 ppl
N_A = np.round(N*.968) ## First face observed response probability is .968
c = 4
w = np.array([.19, .12, .25, .45])

stim = pd.read_csv('faceStim.csv', header = None)
    
exemplars = {
    "A": stim[:5],
    "B": stim[5:10]
}
## len(exemplars)

preds = GCMpred(stim.iloc[0].to_numpy(), exemplars, c, w)
binom.pmf(k = N_A, n = N, p = preds[0])
## Out[163]: 0.02503622303456052

File: ch4/listing4.2.R
pkg <- c("dfoptim", "ggplot2", "data.table")
sapply(pkg, require, character.only = TRUE)
wk <- ifelse(.Platform$OS.type =="windows",
             shortPathName("E:/PPT/ComputationalModelling/ch4/"),
             "/media/yslin/Tui/PPT/ComputationalModelling/ch4/")
setwd(wk)
rm(list = ls())
source("GCMpred.R")

## The data were processed across responses and participants
## There were 80 subjects in the classification experiment ... (see listing4.4)
N <- 2*80  ## two responses per face from 80 ppl
N_A <- round(N*.968)  ## First face observed response probability is .968
N_A

c <- 4
w <- c(.19, .12, .25, .45)

## Table A1:
## Physical specifications and multidimensional scaling (MDS) coordinates for 
## the 34 faces. 
## Eye heights, eye separation, nose length and mouth height
stim <- as.matrix(read.table("faceStim.csv", sep = ","))
exemplars <- list(a = stim[1:5, ], b = stim[6:10, ])
str(stim)
##  num [1:34, 1:4] -1.025 -0.172 -0.98 -0.951 -0.96 ...
##  - attr(*, "dimnames")=List of 2
##   ..$ : NULL
##   ..$ : chr [1:4] "V1" "V2" "V3" "V4"

## GCMpred returns the probability of categorising the 1st face as category A
## and that of category B
preds <- GCMpred(stim[1,], exemplars, c, w)

round(preds, 2)
## [1] 0.93 0.07
likelihood <- dbinom(N_A, size = N, prob = preds[1]); likelihood
likelihood
